### PR TITLE
Pull request

### PR DESCRIPTION
**Pull request by mkaplon**

### CODE REVIEW

File: practice/1_python_part_1/task3.py [hunk 1]
```diff
@@ -16,4 +16,23 @@
 
 
 def build_from_unique_words(*lines: Iterable[str], word_number: int) -> str:
-    ...
+    if (word_number>=0):
+        #text lines into lists
+        a=[i.split(" ") for i in lines]
```
Comment (mkrolczykG) [inline practice/1_python_part_1/task3.py]: " " not needed, default separator in split() func is whitespace

File: practice/1_python_part_1/task4.py [hunk 2]
```diff
@@ -10,4 +10,12 @@
 
 
 def calculate_power_with_difference(ints: List[int]) -> List[int]:
-    ...
+    #squares
+    output=[i**2 for i in ints]
+    for index in range(len(output)):
+        if index>0:
+            output[index]-=output[index-1]-ints[index-1]
```
Comment (mkrolczykG) [inline practice/1_python_part_1/task4.py]: change to **range(1, len(output))** and remove **if**

File: practice/1_python_part_1/task4.py [hunk 3]
```diff
@@ -10,4 +10,12 @@
...
+    return(output)
```
Comment (mkrolczykG) [inline practice/1_python_part_1/task4.py]: Remove redundant parentheses in line 18

File: practice/1_python_part_1/task5.py [hunk 4]
```diff
@@ -13,4 +13,14 @@
 
 
 def remove_duplicated_words(line: str) -> str:
-    ...
+    line_list = line.split(' ')
```
Comment (mkrolczykG) [inline practice/1_python_part_1/task5.py]: remove ' '

File: practice/2_python_part_2/task_read_write.py [hunk 5]
```diff
@@ -13,3 +13,17 @@
 
     result.txt(content: "23, 78, 3")
 """
+import os
+
+def read_files(directory: str):
+    os.chdir(directory)
+    all_values = []
+    for i in range(1, 21):
```
Comment (mkrolczykG) [inline practice/2_python_part_2/task_read_write.py]: Try to change the approach to be more flexible (to be able to handle any number of files in given directory)

File: practice/3_python_testing/tests/test_read_write.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+"""
+Write tests for 2_python_part_2/task_read_write.py task.
+To write files during tests use temporary files:
+https://docs.python.org/3/library/tempfile.html
+https://docs.pytest.org/en/6.2.x/tmpdir.html
+"""
+
+from part2.task_read_write import read_files
```
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_read_write.py:8]: Unresolved reference to function read_files (fix import)

File: practice/3_python_testing/tests/test_read_write.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        # creating files
+        for i in range(1, len(numbers)+1):
+            with open("file_"+str(i)+".txt", 'w') as file:
+                file.write(str(numbers[i-1]))
+        # function which we test
+        read_files(str(tmpdir))
+        # readind output from results.txt
+        with open("result.txt", 'r') as file2:
+            output = file2.read()
+    assert output == expected
```
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_read_write.py]: Isn't better to use **unittest** module for tests?
https://docs.python.org/3/library/unittest.html

File: practice/3_python_testing/tests/test_read_write.py (hunk 6 above)
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_read_write.py]: Unresolved reference to function read_files (fix import)

File: practice/3_python_testing/tests/test_read_write_2.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+"""
+Write tests for 2_python_part_2/task_read_write_2.py task.
+To write files during tests use temporary files:
+https://docs.python.org/3/library/tempfile.html
+https://docs.pytest.org/en/6.2.x/tmpdir.html
+"""
+
+from part2.task_read_write_2 import create_files
```
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_read_write_2.py]: the same as above, "Unresolved reference part2"

File: practice/3_python_testing/tests/test_task_classes.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
+"""
+Write tests for classes in 2_python_part_2/task_classes.py (Homework, Teacher, Student).
+Check if all methods working correctly.
+Also check corner-cases, for example if homework number of days is negative.
+"""
+
+from datetime import datetime, timedelta
+from part2.task_classes import Homework, Student, Teacher
```
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_task_classes.py]: The same as above

File: practice/3_python_testing/tests/test_task_exceptions.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+In case (1,1) it should check if exception were raised
+In case (1,0) it should check if return value is None and "Division by 0" printed
+If other cases it should check if division is correct
+
+TIP: to test output of print() function use capfd fixture
+https://stackoverflow.com/a/20507769
+"""
+
+import pytest, unittest
+from part2.task_exceptions import division, CustomException
```
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_task_exceptions.py]: Same as above

File: practice/3_python_testing/tests/test_task_input_output.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+Tip: for passing custom values to the input() function
+Use unittest.mock patch function
+https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch
+
+TIP: for testing builtin input() function create another function which return input() and mock returned value
+"""
+from unittest.mock import Mock
+from unittest.mock import patch
+from part2.task_input_output import read_numbers
```
Comment (mkrolczykG) [inline practice/3_python_testing/tests/test_task_input_output.py]: Same as above

**Review by mkrolczykG — CHANGES_REQUESTED** [review]
update the things from comments

**Review by mkrolczykG — APPROVED** [review]
